src/state.rs

```diff
@@ -75,14 +75,28 @@ where
     }
 
     /// Deploys a contract to the state, returns the address of the created
-    /// contract or an error
+    /// contract or an error.
+    /// Before the deployment happens the contract's bytecode is instrumented
+    /// and then stored into the NetworkState
     pub fn deploy(
         &mut self,
-        contract: Contract,
-    ) -> Result<ContractId, S::Error> {
-        let id: ContractId = S::Ident::from_bytes(contract.bytecode()).into();
+        mut contract: Contract,
+    ) -> Result<ContractId, VMError<S>> {
+        // Before any instrumentation is applied, generate the contract id.
+        let id: ContractId =
+            S::Ident::from_bytes(contract.code.as_ref()).into();
+
+        // Instrumentalize the contract.
+        ContractInstrumenter::instrument(
+            &mut contract,
+            &crate::Schedule::default(),
+        )?;
```

Comment:
Basically you're using the struct as a function.
You could just have:
```rust
instrument(&mut contract)?;
```
(Since you get the default `Schedule` from the crate you don't need even to pass it).

More, since you have a function that takes a mutable contract, it could be directly a *method* of a contract:
```rust
contract.instrument()?;
```
You could also return a *new* contract instead of mutating the previous one or returns `self` anyway so that you can have:
```rust
.insert(id.clone(), contract.instrument()?)
```

That said, if you instead want to keep the `ContractInstrumentalizer` as public builder API, then you should have *here* something like:
```rust

// you can decide if you want to pass the `Schedule` or instantiate the default internally
let instrumenter = ContractInstrumenter::new(&mut contract)
                                  .with_grow_cost() // this is a wrapper to `ruleset`, that you create in `new`
                                  .with_forbidden_floats() // same here
                                  .with_table_size_limit() // same here for the Schedule
                                  .with_gas_metering()
                                  .with_stack_metering();

instrumenter.inject()?.validate()?;
```

Basically the point is: or you encapsulate everything (and in that case you don't need the structure to be "known" from the outside world in terms of API), or you don't, and let the consumer the decision about how the instrumentalization should be done (with or without floats, with or without gas metering, etc).